### PR TITLE
Feature: "Load Scenario" title sequence command

### PR DESCRIPTION
**Pull request by trigger-death**
**This should not be merged until after #6592**

New command goes by `LOADSC` in script files and in the enumeration.
Scenarios are stored as their filename.
Scenarios selected can only be scenarios to originally come with one of
the games or expansions.
Modified Scenario Select window to have a mode just for title editor
scenario selection.

### Strings Added:
```
STR_TITLE_EDITOR_ACTION_LOAD_SAVE = 6159,
STR_TITLE_EDITOR_ACTION_LOAD_SCENARIO = 6160,
STR_TITLE_EDITOR_ARGUMENT_SCENARIO = 6161,
STR_TITLE_COMMAND_EDITOR_NO_SCENARIO_SELECTED = 6162,
STR_TITLE_EDITOR_COMMAND_LOAD_NO_SCENARIO = 6163,
STR_TITLE_COMMAND_EDITOR_ACTION_SELECT_SCENARIO = 6164,
```

### Preview

Preview

### Future of RCT1/RCT2 Title Sequences

Once this feature is merged. The RCT1/RCT2 title sequences will no longer need to rely on the `LOADMM` and `LOADRCT1` commands. You guys can decide if it's worth keeping these commands afterwords or not. Either way, this switch will make duplicating and editing existing sequences more user-friendly. Of course we shouldn't bother modifying the current title sequences until it's time for the next stable release.

**Edit:** Alternatively we can have the script reader automatically convert `LOADMM` and `LOADRCT1` commands into `LOADSC` commands during parsing in order to remove the need for their support.

### CODE REVIEW

File: src/openrct2-ui/windows/TitleEditor.cpp [hunk 1]
```diff
@@ -936,6 +936,26 @@ static void window_title_editor_scrollpaint_commands(rct_window *w, rct_drawpixe
...
+            const char * name = "";
+            source_desc desc;
+            if (command->Scenario[0] == '\0')
+            {
+                commandName = STR_TITLE_EDITOR_COMMAND_LOAD_NO_SCENARIO;
+            }
+            else
+            {
+                auto scenario = GetScenarioRepository()->GetByFilename(command->Scenario);
+                if (scenario != nullptr && scenario_get_source_desc(scenario->name, &desc))
```
Comment (janisozaur) [inline src/openrct2-ui/windows/TitleEditor.cpp]: `ScenarioSources::TryGetByName`?

File: src/openrct2/title/TitleSequence.h [hunk 2]
```diff
@@ -32,6 +32,7 @@ typedef struct TitleCommand
         uint8 Zoom;         // ZOOM
         uint8 Speed;        // SPEED
         uint16 Milliseconds;      // WAIT
+        utf8 Scenario[64];        // LOADSC
```
Comment (janisozaur) [inline src/openrct2/title/TitleSequence.h]: Is 64 arbitrary or some limit?

File: src/openrct2-ui/windows/TitleEditor.cpp (hunk 1 above)
Comment (trigger-death) [reply src/openrct2-ui/windows/TitleEditor.cpp]: `TryGetByName` attempts to locate the title sequence based on it's translated title. This won't work when sharing title sequences between different languages.

File: src/openrct2/title/TitleSequence.h (hunk 2 above)
Comment (trigger-death) [reply src/openrct2/title/TitleSequence.h]: 64 is arbitrary, but also what is used in other areas. Currently the longest built-in title sequence name is 50 characters long exluding the null character at the end.

File: src/openrct2-ui/windows/TitleEditor.cpp (hunk 1 above)
Comment (janisozaur) [reply src/openrct2-ui/windows/TitleEditor.cpp]: https://github.com/OpenRCT2/OpenRCT2/blob/00fd18809ca71e4b73e8d3c4d658d2139d5dd881/src/openrct2/scenario/ScenarioSources.cpp#L361-L364

Comment (trigger-death) [reply src/openrct2-ui/windows/TitleEditor.cpp]: It looks like it won't work if I use it based on the scenario name even if it isn't language based. At the moment, there's no function to convert from scenario name to scenario path. That and scenarios in RCT1 are named sc1.SC4 and not their actual name. Again, it stores the scenario as it's filepath.

File: src/openrct2/title/TitleSequence.h (hunk 2 above)
Comment (rwjuk) [reply src/openrct2/title/TitleSequence.h]: We should have a `#define` for this limit then.

**Comment by AaronVanGeffen** [comment]
I have a question regarding the UI… In the 'Command Editor' window, the 'Scenario to load' is printed in what appears to be an input box. However, it does not seem to allow manually entering any text there; instead, we have to use the 'Select' button. Therefore, I believe it would be better to do away with the input box frame altogether, and instead just draw the string as-is.

On a sidenote, I can't seem to duplicate the OpenRCT2 title sequence ("contains invalid data"). Other built-in sequences seem fine, however.

**Comment by trigger-death** [comment]
Personally I think the frame for the scenario helps separate that it's the argument for the command. But if other people agree that it should be hidden then I'm fine with that.

I'll look into the title sequence dupe issue.

File: src/openrct2/title/TitleSequence.h (hunk 2 above)
Comment (trigger-death) [reply src/openrct2/title/TitleSequence.h]: Added

Comment (AaronVanGeffen) [reply src/openrct2/title/TitleSequence.h]: It's nice you've added the `#define`, but you're still hardcoding the 64 on line 37. :wink:

Comment (trigger-death) [reply src/openrct2/title/TitleSequence.h]: My mistake. I swear I added that in there. Ah well.

**Comment by Gymnasiast** [comment]
Can you rebase this?

Also, it should take the names as listed in the translation files - meaning that if you have your game set to US English and you select "Aqua Park", it must be listed as "White Water Park" in the file. Does it do this?

Finally, after this is merged, we'll adjust the title sequence scripts to remove the old commands. After that, we can cut them from the code.

**Comment by trigger-death** [comment]
All rebased. And yes, the filename is what's listed in the script file. The translated names are only displayed in the UI.

**Comment by Gymnasiast** [comment]
Tested it.

- I can't select WW scenarios like 'Cradle of Civilisation'.
- This will not work universally for RCT1 scenarios. We use an identifier like `White Water Park` internally, not a filename, since that can be `sc9.sc4`, `White Water Park.SC6` or `Aqua Park.SC6`.

**Comment by trigger-death** [comment]
Okay I see what you mean. I'll get on that.

**Comment by Gymnasiast** [comment]
With your changes, I can now load TT scenarios, but no RCT1 scenarios.

**Comment by trigger-death** [comment]
Fixed the issue. I didn't notice the scenario index repository had a different function to create sc4 entries. As such, you'll need to delete your scenarios.idx file if the last built version was from the previous commit.

**Comment by Gymnasiast** [comment]
This does not compile because it can't find util.h. Remember that Linux is case-sensitive.

**Comment by trigger-death** [comment]
Would you be able to look into that? I don't have a Linux environment to test on.

**Comment by Gymnasiast** [comment]
The crash I have does not happen with develop. I haven't talked about missing, I didn't have that bug yesterday.

**Comment by trigger-death** [comment]
I misread your last comment. I thought I needed to change it to "util.h" in the include because it was already correctly labeled as "Util.h"

**Comment by AaronVanGeffen** [comment]
I've taken the liberty of rebasing this PR. When testing, my title sequence is able to load both SV4 and SV6 scenarios, from both the base game and their expansions. Code-wise, I've only had a quick look, but don't see any objections.

@Gymnasiast, could you give this one another go as well?

**Comment by Gymnasiast** [comment]
I will.

File: data/language/en-GB.txt [hunk 3]
```diff
@@ -1173,8 +1173,8 @@ STR_1167    :Can't raise water level here...
 STR_1168    :Options
 STR_1169    :(None)
 STR_1170    :{STRING}
-STR_1171    :{RED}Closed - - 
-STR_1172    :{YELLOW}{STRINGID} - - 
+STR_1171    :{RED}Closed - -
```
Comment (Gymnasiast) [inline data/language/en-GB.txt]: Please revert these whitespace changes.

**Comment by trigger-death** [comment]
@Gymnasiast and @AaronVanGeffen I appreciate the help finishing off this and the follow sprite PR in my absence. Glad to finally see it merged.

**Pull request merged**